0.15 features

Foxyas opened the pull request:
Updated pt_br
Added Lang Keys for new attributes for en_us
Fixed #1538 for 1.20.1
Fixed #1545 for 1.20.1
Fixed #1552 for 1.20.1
Fixed Mixin compatibility.Forgery.LatexHumanoidArmorLayerMixin using old renderModel method.
 *Side Note: it didn't stop me from compiling the mod, whoever it is something to fix? idk for sure soo i just let the old code commented and the "new code" applied down below*
Fixed Weird "ride" animation when laying down into a bed with the mod "sleep tight"
Fixed ColorSelector$renderWidget not rendering the colors
 *it was using 0x00colorInt instead of 0xffcolorint*
Fixed Breathing Bug where the player would lost air even with Water Breathing Effect. 
*also added a simple check for compatibility.. who knows right?*

Create a Temp mixin to fix the accesstransformer not applying the "remove final" from FluidType$isAir
Added some datagens to help with the dev
  - to use just call add (or similar) in the file and then run the task "runData"
  - the folder where it stores the generated files are `generated/resources`
  - Embedium or any other client side mod will crash the runData task so be careful

publish w.i.p version into maven to help addon dev's :>

## Proposed Future Changes (Need approval from main dev first)
- [ ] - Add Interface ICrawlEntity 
    - will make any changed entity able to have the "safe pose check" feature similar to the player
    - will make any changed entity also able to enter into "crawl mode" if they target is in crawl pose
- [ ] - Add Interface IGrabberEntity
    - will make any changed entity able to use the grab ability
    - gonna be a dependence for "grab target goals"
- [ ] - Add a ´/accessory´ command that gonna allow any admin to set a accessory for any entity/entities
    - Dynamic slot suggestion system (only show the slots that the entity have and if something bad happen that the code return a invalid state exception will just show all the slots registred)
    - Dynamic FeedBack to tell the command executor what wrong happen
- [ ] - Add Face/Head Accessory Slot
    - For the Income features :>

Foxyas committed:
created a maven with all the stuff needed

Foxyas committed:
higbgu7

Foxyas committed:
Update ChunkRenderDispatcherMixin.java

Foxyas committed:
Update ChunkRenderDispatcherMixin.java

Foxyas committed:
Update ChunkRenderDispatcherMixin.java

manually did a check for indexs

Foxyas committed:
tweaks

Foxyas committed:
fixed issues and added datagens

fixed #1545
fixed the dupe glitch of #1538

Foxyas committed:
tweaks

Foxyas committed:
Refactor latex cover state handling for thread safety

Replaces the static latexCoverStateGetter with a ThreadLocal to ensure thread safety in LatexCoveredBlocksRenderer. Removes old fail-safe code in ChunkRenderDispatcherMixin and updates related usages for improved concurrency handling.

Foxyas committed:
published the version in maven

Foxyas committed:
Update ColorSelector.java

fixed wrong alpha into colors

Foxyas committed:
published the new version into maven :>

Foxyas committed:
Add new attributes and update translations

Introduced new attributes (step size, sprint speed, sneak speed, air capacity) to en_us.json and pt_br.json. Improved and expanded the Brazilian Portuguese translation for items, blocks, abilities, subtitles, and commands. Fixed logic in TransfurVariantInstance to prevent air supply reduction for entities with water breathing or that cannot drown.

Foxyas committed:
maven publish :>

Foxyas committed:
Add Sleep Tight mod compatibility for ride animators

Introduced mixins to support the Sleep Tight mod by cancelling ride animation setup for entities riding BedEntity. Updated build.gradle to include Sleep Tight as a dependency and registered new mixins in changed.compatibility.mixins.json.

Foxyas committed:
Update hideArmor injection mthod signature

Commented out the old hideArmor method and added a new @Inject with an updated method signature to match changes in parameters. This resolves compiler warnings and ensures compatibility with the latest codebase.

Foxyas committed:
Update comment in LatexHumanoidArmorLayerMixin

Clarified the comment regarding compiler warnings and patch application location in LatexHumanoidArmorLayerMixin.java.

Foxyas committed:
Final Tweaks and polishing

Fixed offset when "riding" the bedEntity
Fixed Arms being lifted when "riding" the bedEntity

added new file cuz moonLightLib don't work with curseMaven

Foxyas committed:
published maven for addons :>

Foxyas committed:
Update BedEntityMixin.java

Foxyas committed:
maven :>

Foxyas committed:
refreshing maven

LtxProgrammer commented on src/main/java/net/ltxprogrammer/changed/client/LatexCoveredBlocksRenderer.java:
```diff
@@ -377,10 +373,13 @@ public RenderType getRenderType(LatexCoverState coverState) {
-    private static LatexCoverGetter latexCoverStateGetter = null;
+//    Old Code
+//    @Nullable
+//    private static LatexCoverGetter latexCoverStateGetter = null;
```
No need to leave old code in a comment, the diff on git is enough.

Foxyas replied:
i tweaked

LtxProgrammer commented on src/main/java/net/ltxprogrammer/changed/client/LatexCoveredBlocksRenderer.java:
```diff
@@ -480,7 +483,7 @@ public boolean tesselate(
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Tesselating latex cover in world");
             CrashReportCategory crashreportcategory = crashreport.addCategory("Block being tesselated");
-            CrashReportCategory.populateBlockDetails(crashreportcategory, level, blockPos, (BlockState)null);
+            CrashReportCategory.populateBlockDetails(crashreportcategory, level, blockPos, (BlockState) null);
```
The code base uses C-style casting all over the place. This formatting adjustment is not necessary.

Foxyas replied:
InteliJ auto did it when i did the "optimize imports" thing

Foxyas replied:
i tweaked

LtxProgrammer commented on src/main/java/net/ltxprogrammer/changed/client/LatexCoveredBlocksRenderer.java:
```diff
@@ -527,13 +530,13 @@ private static CompletableFuture<Map<Block, LatexModelDefinition>> loadBlockStat
         Map<Property<?>, Comparable<?>> map = Maps.newHashMap();
 
-        for(String s : COMMA_SPLITTER.split(text)) {
+        for (String s : COMMA_SPLITTER.split(text)) {
```
Unlike the comment on casting, this is a change I do appreciate. This instance was likely a holdover from copying Minecraft's source.

Foxyas replied:
the idea was (@0SeNiA0) idea :p

LtxProgrammer commented on src/main/java/net/ltxprogrammer/changed/entity/variant/TransfurVariantInstance.java:
```diff
@@ -851,11 +852,13 @@ protected void tickBreathing() {
 
-        if (!host.level().isClientSide)
+        if (!host.level().isClientSide) {
+            if (host.hasEffect(MobEffects.WATER_BREATHING) || !host.canDrownInFluidType(ForgeMod.WATER_TYPE.get())) return;
```
This fix is in the wrong place. It should instead be where the code specifically checks breathing underwater (line 805).

Foxyas replied:
Ohh okay, i just thought that tickBreathing was the right place :p

Foxyas replied:
Ohh okay, i just thought that tickBreathing was the right place :p

Foxyas replied:
i tweaked

LtxProgrammer replied:
I pushed a commit to properly handle this case.

LtxProgrammer commented on src/main/java/net/ltxprogrammer/changed/mixin/compatibility/SleepTight/ArmRideAnimatorMixin.java:
Wouldn't this fix be simpler if you instead overwrote the conditions for `HumanoidAnimator.AnimateStage.RIDE` and `HumanoidAnimator.AnimateStage.SLEEP`?

Foxyas replied:
I tried and the animation still happen which looks odd.. i gonna try doing more tweaks

Foxyas replied:
i gonna try again, applying other check now

Foxyas replied:
i gonna apply a mixin into HumanoidAnimator.AnimateStage$test which i probably think that will work

Foxyas replied:
<img width="689" height="436" alt="image" src="https://github.com/user-attachments/assets/41ebf0b9-8627-496b-82f0-5b2c20a1b091" />
it works

Foxyas replied:
<img width="553" height="430" alt="image" src="https://github.com/user-attachments/assets/23c183ab-f741-471b-b9da-7d2b8d7f1eea" />
the "normal sleep"

Foxyas replied:
i need to tweak the offset tho

Foxyas replied:
<img width="393" height="203" alt="image" src="https://github.com/user-attachments/assets/0a521359-2b74-4d09-b405-42f239ca1915" />
i got it working but the hitbox is abit off...
however the hitbox that is out of the bed is the "bed entity" not the player one

Foxyas replied:
@LtxProgrammer should i apply a animation mixin to offset the whole root of the model or just keep it was it is?

LtxProgrammer replied:
What does it look like for a humanoid latex?

Foxyas replied:
> What does it look like for a humanoid latex?

for humanoids the hitbox is still glued to the bed, but still has a bit of offset.
it looks VERY similar to the "true sleeping position" too
<img width="781" height="453" alt="image" src="https://github.com/user-attachments/assets/1cb31bc6-80a7-42e0-bbf8-0faf66256d2a" />

Foxyas replied:
@LtxProgrammer i already made a better handle
<img width="287" height="179" alt="image" src="https://github.com/user-attachments/assets/62dbdc1a-0d75-4af6-a988-2bc2974ab251" />
<img width="356" height="206" alt="image" src="https://github.com/user-attachments/assets/ae90c483-5f59-4102-9a6c-7581fab287e8" />

Foxyas replied:
i still need todo some for the quadrupdals

Foxyas replied:
anyway i gonna do a simple "switch" for the offset

LtxProgrammer commented on src/main/java/net/ltxprogrammer/changed/fluid/Gas.java:
```diff
@@ -41,10 +41,10 @@ public GasFluidType(Properties properties) {
+//        @Override
+//        public boolean isAir() { // Requires accesstransformer to work
+//            return true;
+//        } // For some Reason the Access Transformer stop working? check the FluidTypeMixin
```
This could be a dev env bug. I would need to test myself.

Foxyas replied:
alright i will let that then

LtxProgrammer replied:
I was able to compile and run the mod without the above fix. You may have not run the correct steps to compile after an accesstransformer change (genIntellijRuns).

Foxyas replied:
bruh....

Foxyas replied:
nop.. it's still broken

Foxyas replied:
i got it fixed @LtxProgrammer i just had to update the file with another access transformer

Foxyas replied:
aka i add this line
```java
public net.minecraft.world.damagesource.CombatTracker f_19282_ # takingDamage
```
which gonna be useful for the exoskeleton

LtxProgrammer commented on src/main/java/net/ltxprogrammer/changed/mixin/client/RenderChunkRegionAccessor.java:
Vestigial accessor? I don't see any references to it elsewhere.

Foxyas replied:
is from my old fix i gonna remove it

LtxProgrammer commented on src/main/java/net/ltxprogrammer/changed/mixin/compatibility/Forgery/LatexHumanoidArmorLayerMixin.java:
I'm aware of a few cases where the IDE warns on target methods, but there is no need for the changes present in this file.

Foxyas replied:
i check it again, and there's fr no method with this dependencies
```java
renderModel(T entity, ItemStack stack, EquipmentSlot slot, PoseStack pose, MultiBufferSource buffers, int packedLight, boolean foil, LatexHumanoidArmorModel<T, ?> model, float red, float green, float blue, ResourceLocation armorResource)
```

the most similar is this 
```java
renderModel(T entity, ItemStack stack, EquipmentSlot slot, PoseStack pose, MultiBufferSource buffers, int packedLight, LatexHumanoidArmorModel<? super T, ?> model, float red, float green, float blue, ResourceLocation armorResource)
```

i check again and the old method is from 1.18.2 
this line to be specific -> https://github.com/LtxProgrammer/Changed-Minecraft-Mod/blob/fea4476ca582ecb64e1cb98900e2749ea15dfa4f/src/main/java/net/ltxprogrammer/changed/client/renderer/layers/LatexHumanoidArmorLayer.java#L101

LtxProgrammer commented on src/main/java/net/ltxprogrammer/changed/datagen/ability_tree/AbilityTreeProviderImpl.java:
Currently serves no purpose other than to showcase an example on how to use it.

Foxyas replied:
i removed the class but keep the main datagen thing

LtxProgrammer commented on gradle.properties:
```diff
@@ -7,7 +7,7 @@ org.gradle.daemon=false
 
 runWithClientOnlyMods = true
-runWithMods = embeddium,oculus,presencefootsteps,spark,modernfix,ferritecore,entityculling,debugutils
+runWithMods = embeddium,oculus,presencefootsteps,spark,modernfix,ferritecore,entityculling,debugutils,sleep_tight,moonlight_lib
```
I know I never said this, but the mods that are in this list are intended to be vital in the run environment. Mods that are present in this list only for debugging a compatibility should not be committed.

Foxyas replied:
Oops.. i forgot to remove it :p

Foxyas replied:
Oops.. i forgot to remove it :p

Foxyas replied:
i tweaked

LtxProgrammer commented on src/main/java/net/ltxprogrammer/changed/mixin/client/ChunkRenderDispatcherMixin.java:
```diff
@@ -34,12 +36,20 @@
+
+        //Old Fail Safe
+        //if (i < 0 || i >= region.chunks.length) return ChangedLatexTypes.NONE.get().defaultCoverState();
+        //if (j < 0 || j >= region.chunks[0].length) return ChangedLatexTypes.NONE.get().defaultCoverState();
```
Vestigial commented-out code.

Foxyas replied:
i removed the old code comment :P

LtxProgrammer reviewed (changes requested):
Please see review comments.

Foxyas committed:
Update gradle.properties

removed sleep_tight and moonlight_lib from the graddle.properties

Foxyas committed:
Update LatexCoveredBlocksRenderer.java

is the cast over null is really necessary?
anyway i keep it

Foxyas committed:
removed AbilityTreeProviderImpl

i kept the main datagen for abilities tho, you can use it instead of writing each json file again and again

Foxyas committed:
removed old code comments

Foxyas commented:
@LtxProgrammer do you want to me to remove the maven files?

LtxProgrammer committed:
Corrected underwater breathing logic to consider water breathing effect

Foxyas commented:
oh so it should reset the value. okay

LtxProgrammer commented:
> oh so it should reset the value. okay

`ticksBreathingUnderwater` is to track how long the entity has relied on their own ability to breathe water in order to progress the advancement.

`airDelta` is how much the air will change that tick. Positive adds air, and negative subtracts.

Foxyas commented:
> > oh so it should reset the value. okay
> 
> `ticksBreathingUnderwater` is to track how long the entity has relied on their own ability to breathe water in order to progress the advancement.
> 
> `airDelta` is how much the air will change that tick. Positive adds air, and negative subtracts.

soo the player will gain air with a water breath potion? correct me if i'm wrong but the potion effect does that normaly?

Foxyas commented:
If I remember correctly, the potion prevents loss of air. But if the player drinks the potion after losing some "Breath Ticks," the player will still retain the same breath ticks without recovering them. Which I think doesn't make much sense in a certain way.
"Water Breathing"

LtxProgrammer commented:
> @LtxProgrammer do you want to me to remove the maven files?

My only concern with the maven files is the existence of a published copy of the jar.

> soo the player will gain air with a water breath potion? correct me if i'm wrong but the potion effect does that normaly?

You are correct, I hastily modified the code without reviewing vanilla behavior. I don't know why Mojang named it "water breathing" if the player isn't "breathing" the water. I'll revert the change.

LtxProgrammer committed:
Change airDelta value for water breathing potion effect

Foxyas commented:
> > @LtxProgrammer do you want to me to remove the maven files?
> 
> My only concern with the maven files is the existence of a published copy of the jar.
> 
> > soo the player will gain air with a water breath potion? correct me if i'm wrong but the potion effect does that normaly?
> 
> You are correct, I hastily modified the code without reviewing vanilla behavior. I don't know why Mojang named it "water breathing" if the player isn't "breathing" the water. I'll revert the change.

yeah mojang really need to redo some features logic LOL

Foxyas committed:
fixed access transformer

Foxyas committed:
Update ExoskeletonItem.java

Foxyas committed:
Update Gas.java

removed old code comment

LtxProgrammer commented on src/main/java/net/ltxprogrammer/changed/item/ExoskeletonItem.java:
This change is no longer necessary.

Foxyas replied:
okay

Foxyas committed:
Update ExoskeletonItem.java

Foxyas commented:
@LtxProgrammer do you wanna for me to commit the /accessory command?

LtxProgrammer commented:
> @LtxProgrammer do you wanna for me to commit the /accessory command?

If you can do it well, but only after resolving my comment on the sleep tight mod.

LtxProgrammer committed:
Merge branch '0.15-features' into 0.15-features

Foxyas committed:
finally the final tweaks

Foxyas commented:
@LtxProgrammer i probably made all the needed changes into the compatibility with sleep Tight

it now uses the livingRenderer setUpRotations instead of the "rider position" soo the hitbox is right

i made a offset multiplier for each entity shape instead of entity instance types

Foxyas commented:
The accessory item thing will probably take a while to be commited cuz a heavy storm is reaching my region and i really don't want to risk using my pc while that.

LtxProgrammer commented:
> The accessory item thing will probably take a while to be commited cuz a heavy storm is reaching my region and i really don't want to risk using my pc while that.

Very well, it can be in another PR. What are you worried about with the storm, an electric surge? You should get a surge protector. Or a UPS if its a blackout you're worried about.

Foxyas commented:
@LtxProgrammer btw, would you allow me to add the grab feature for entities?

I can make where entities only try to grab the player if they are inside of a "can grab" tag.

Check changed addon repository for examples

Check the interface "IGrabberEntity"

Foxyas commented:
> > The accessory item thing will probably take a while to be commited cuz a heavy storm is reaching my region and i really don't want to risk using my pc while that.
> 
> Very well, it can be in another PR. What are you worried about with the storm, an electric surge? You should get a surge protector. Or a UPS if its a blackout you're worried about.

Yeah basically that. And yeap i gonna buy one when i got some money :p
